rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,324 @@
+sections, `Any` requires assumptions on the underlying types, it is the case of
+trait reflection as well.
+
+The problem stems from some methods of old `dyn Reflect` assuming that the
+underlying type is always the same. 
+
+The following methods are the one that assumes uniqueness:
+- `any`
+- `any_mut`
+- `downcast`
+- `take`
+- `is`
+- `downcast_ref`
+- `downcast_mut`
+- `set`
+
+The `Any` trait bound on old `Reflect` is similarly problematic, as it introduces
+the same issues as those methods.
+
+Note that the following methods are also problematic, and require discussion,
+but to keep to scope of this RFC short, we will keep them in `PartialReflect`.
+- `reflect_hash`
+- `reflect_partial_eq`
+- `serializable`
+
+## Proposition
+
+This RFC separates the bits of `Reflect` needed for (1) and the bits needed for
+(2) and (3).
+
+We introduce `PartialReflect` to isolate the methods of `Reflect` for (1). This
+also means that the old `Reflect` subtraits now are bound to `PartialReflect`
+rather than `Reflect`. Proxy types such as `DynamicStruct` will also now only
+implement `PartialReflect` instead of full `Reflect`, because they do not make
+sense in the context of (2) and (3).
+
+This makes `Reflect` "canonical", in that the underlying type is the one being
+described all the time. And now the `Any` bound and any-related methods are all
+"safe" to use, because the assumptions requires to be able to use them are
+backed into the type system.
+
+We still probably want a way to convert from a `PartialReflect` to a `Reflect`,
+so we add a `into_full` method to `PartialReflect` to convert them into
+`Reflect` in cases where the underlying type is canonical.
+
+The `FromReflect` trait will be renamed `FromPartialReflect`, a future
+implementation might also add the ability to go from any _proxy_ partial
+reflect into a full reflect using a combination of `FromPartialReflect` and the
+type registry.
+
+In short:
+- create a new trait: `PartialReflect`
+- `Reflect: PartialReflect`
+- All methods on `Reflect` that **does not** depend on the uniqueness of the
+  underlying type are moved to `PartialReflect`
+- Most things should now accept a `dyn PartialReflect` rather than
+  `dyn Reflect`.
+- It is possible to convert a `PartialReflect` into a `Reflect` using a
+  `into_full` method.
+- `FromReflect` becomes `FromPartialReflect`.
+
+
+### `PartialReflect` trait
+
+All methods that do not assume underlying uniqueness should be moved to a new
+trait: `PartialReflect`. `PartialReflect` also has a `as_full` and `into_full`
+to convert them into "full" `Reflect`. We also keep the "trait reflection"
+methods, because changing the trait reflection design is complex and we want to
+keep the scope of this RFC minimal.
+
+```rust
+pub trait PartialReflect: Send + Sync {
+    // new
+    fn as_full(&self) -> Option<&dyn Reflect>;
+    fn into_full(self: Box<Self>) -> Option<Box<dyn Reflect>>;
```

Comment:
I think this should return `Result<Box<dyn Reflect>, Box<dyn PartialReflect>>` to allow getting back the original `PartialReflect` in case `Reflect` is not implemented.